management-server/internal/xds/server.go

```diff
@@ -81,16 +85,175 @@ func FeedData() {
+	version := rand.Intn(maxRandomInt)
+
+	for _, event := range applicationEventArray {
+		label := event.Label
+		appUUID := event.UUID
+
+		application, err := database.GetApplicationByUUID(appUUID)
+		if err != nil {
+			logger.LoggerDatabase.ErrorC(logging.ErrorDetails{
+				Message: fmt.Sprintf("Error retrieving application for uuid : %s from database error: %v, "+
+					"hence skipping add to xdx cache", appUUID, err),
+				Severity:  logging.MINOR,
+				ErrorCode: 1101,
+			})
+			continue
+		}
+
+		snapshotEntry, snapshotFound := snapshotMap[label]
+		var newSnapshot wso2_cache.Snapshot
+
+		if !snapshotFound {
+			newSnapshot, _ = wso2_cache.NewSnapshot(fmt.Sprint(version), map[wso2_resource.Type][]types.Resource{
+				wso2_resource.APKMgtApplicationType: {application},
+			})
+			snapshotEntry = &newSnapshot
+			snapshotMap[label] = &newSnapshot
+		} else {
+			// error occurs if no snapshot is under the provided label
+			resourceMap := snapshotEntry.GetResourcesAndTTL(typeURL)
+			resourceMap[appUUID] = types.ResourceWithTTL{
+				Resource: application,
+			}
+			appResources := convertResourceMapToArray(resourceMap)
+			newSnapshot, _ = wso2_cache.NewSnapshot(fmt.Sprint(version), map[wso2_resource.Type][]types.Resource{
+				wso2_resource.APKMgtApplicationType: appResources,
+			})
+			snapshotMap[label] = &newSnapshot
+		}
+	}
+	applicationCacheMutex.Lock()
+	defer applicationCacheMutex.Unlock()
+	for label, snapshotEntry := range snapshotMap {
+		applicationCache.SetSnapshot(context.Background(), label, *snapshotEntry)
+		introducedLabels[label] = true
+		logger.LoggerXds.Infof("Application Snaphsot is updated for label %s with the version %d.", label, version)
+	}
+}
+
+func convertResourceMapToArray(resourceMap map[string]types.ResourceWithTTL) []types.Resource {
+	var appResources []types.Resource
+	for _, res := range resourceMap {
+		appResources = append(appResources, res.Resource)
+	}
+	return appResources
+}
+
+// SetEmptySnapshot sets an empty snapshot into the applicationCache for the given label
+// this is used to set empty snapshot when there are no Applications available for a label
+func SetEmptySnapshot(label string) error {
+	version := rand.Intn(maxRandomInt)
+	newSnapshot, err := wso2_cache.NewSnapshot(fmt.Sprint(version), map[wso2_resource.Type][]types.Resource{
+		wso2_resource.APKMgtApplicationType: {},
+	})
+	if err != nil {
+		logger.LoggerXds.Errorf("Error creating empty snapshot. error: %v", err.Error())
+		return err
+	}
+	applicationCacheMutex.Lock()
+	defer applicationCacheMutex.Unlock()
+	//performing null check again to avoid race conditions
+	_, errSnap := applicationCache.GetSnapshot(label)
+	if errSnap != nil && strings.Contains(errSnap.Error(), "no snapshot found for node") {
+		errSetSnap := applicationCache.SetSnapshot(context.Background(), label, newSnapshot)
+		if errSetSnap != nil {
+			logger.LoggerXds.Errorf("Error setting empty snapshot to apiCache. error : %v", errSetSnap.Error())
```

Comment:
follow proper logs with error codes

Reply:
fixed in https://github.com/wso2/apk/pull/151/commits/4f4c37b97abe8646e783d357a3998601be6de56c